## 1、归一化，标准化，缺失值插补练习

### 1.1 归一化

In [1]:
import numpy as np
from sklearn.preprocessing import MinMaxScaler, StandardScaler

# 创建数据
data = np.array([[1,200],[2,300],[3,400]])
print("原始数据：")
print(data)

# Min-Max归一化（缩放到[0,1]范围）
scaler = MinMaxScaler()  # 默认缩放区间[0,1]
data_normalized = scaler.fit_transform(data)
print("\nMin-Max归一化后：")
print(data_normalized)

原始数据：
[[  1 200]
 [  2 300]
 [  3 400]]

Min-Max归一化后：
[[0.  0. ]
 [0.5 0.5]
 [1.  1. ]]


### 1.2 标准化

In [2]:
import numpy as np
from sklearn.preprocessing import StandardScaler

# 创建数据
data = np.array([[1,200],[2,300],[3,400]])
print("原始数据：")
print(data)

# Z-score标准化（均值为0，标准差为1）
scaler = StandardScaler()
data_standardized = scaler.fit_transform(data)
print("\nZ-score标准化后：", data_standardized)
print("\n标准化后的均值：", data_standardized.mean(axis=0))
print("标准化后的标准差：", data_standardized.std(axis=0))

原始数据：
[[  1 200]
 [  2 300]
 [  3 400]]

Z-score标准化后： [[-1.22474487 -1.22474487]
 [ 0.          0.        ]
 [ 1.22474487  1.22474487]]

标准化后的均值： [0. 0.]
标准化后的标准差： [1. 1.]


### 1.3 缺失值处理

In [3]:
import numpy as np
from sklearn.impute import SimpleImputer

# 创建包含缺失值的数据
data = np.array([[1, np.nan], [2, 300], [np.nan, 400]])
print("原始数据（包含缺失值）：")
print(data)

# 使用均值填充缺失值（默认填充策略）
imputer_mean = SimpleImputer(strategy='mean')
data_imputed_mean = imputer_mean.fit_transform(data)
print("\n使用均值填充后：")
print(data_imputed_mean)

# 使用中位数填充缺失值
imputer_median = SimpleImputer(strategy='median')
data_imputed_median = imputer_median.fit_transform(data)
print("\n使用中位数填充后：")
print(data_imputed_median)

# 使用众数填充缺失值
imputer_mode = SimpleImputer(strategy='most_frequent')
data_imputed_mode = imputer_mode.fit_transform(data)
print("\n使用众数填充后：")
print(data_imputed_mode)

原始数据（包含缺失值）：
[[  1.  nan]
 [  2. 300.]
 [ nan 400.]]

使用均值填充后：
[[  1.  350. ]
 [  2.  300. ]
 [  1.5 400. ]]

使用中位数填充后：
[[  1.  350. ]
 [  2.  300. ]
 [  1.5 400. ]]

使用众数填充后：
[[  1. 300.]
 [  2. 300.]
 [  1. 400.]]


import numpy as np
from sklearn.decomposition import PCA

# 创建数据
X = np.array([[0, 2, 0, 3],
              [0, 1, 4, 3],
              [0, 1, 1, 3]])
print("原始数据 X：")
print(X)
print("原始数据形状：", X.shape)

# 指定累计解释方差95%进行PCA降维
pca_95 = PCA(n_components=0.95)  # 保留累计解释方差≥95%的主成分
X_pca_95 = pca_95.fit_transform(X)
print("\n使用累计解释方差95%进行PCA降维：")
print("降维后的数据：")
print(X_pca_95)
print("降维后的数据形状：", X_pca_95.shape)
print("\n各主成分解释的方差比例：", pca_95.explained_variance_ratio_)
print("累计解释的方差比例：", pca_95.explained_variance_ratio_.cumsum())
print(f"实际保留的主成分数量：{pca_95.n_components_}")

### 2.1 特征选择：方差阈值

移除方差极小的特征

In [4]:
import numpy as np
from sklearn.feature_selection import VarianceThreshold

# 创建数据
X = np.array([[0, 2, 0, 3],
              [0, 1, 4, 3],
              [0, 1, 1, 3]])
print("原始数据 X：")
print(X)

# 计算各列的方差
variances = np.var(X, axis=0)
print("\n各列的方差：", variances)

# 默认阈值为0（移除方差<= 0 的特征）
selector = VarianceThreshold()
X_selected = selector.fit_transform(X)
print("\n使用VarianceThreshold（threshold=0）后：")
print(X_selected)
print("保留的特征索引：", selector.get_support())

# 自定义阈值：移除方差<= 0.5 的特征
selector_custom = VarianceThreshold(threshold=0.5)
X_selected_custom = selector_custom.fit_transform(X)
print("\n使用VarianceThreshold（threshold=0.5）后：")
print(X_selected_custom)
print("保留的特征索引：", selector_custom.get_support())

原始数据 X：
[[0 2 0 3]
 [0 1 4 3]
 [0 1 1 3]]

各列的方差： [0.         0.22222222 2.88888889 0.        ]

使用VarianceThreshold（threshold=0）后：
[[2 0]
 [1 4]
 [1 1]]
保留的特征索引： [False  True  True False]

使用VarianceThreshold（threshold=0.5）后：
[[0]
 [4]
 [1]]
保留的特征索引： [False False  True False]


### 2.2 PCA 降维


**主成分解释的方差比例**

- 方差 = 数据信息量
- 解释的方差比例 = 某主成分的方差 ÷ 总方差
- 含义：该主成分能**保留原始数据多少信息**

**要点：**

- 主成分按解释方差从大到小排序
- 常用累计解释方差 ≥ 90% / 95% 来决定保留的维度

**sklearn：**

```python
pca.explained_variance_ratio_      # 各主成分比例
pca.explained_variance_ratio_.cumsum()  # 累计比例
```

**一句话**：
✅解释的方差比例越大，该主成分越重要。

In [5]:
import numpy as np
from sklearn.decomposition import PCA

# 创建数据
X = np.array([[0, 2, 0, 3],
              [0, 1, 4, 3],
              [0, 1, 1, 3]])
print("原始数据 X：")
print(X)
print("原始数据形状：", X.shape)

print('-'*50)

# 使用PCA进行降维
# n_components 是整数时：保留的维度
# n_components 是小数 (0~1) 时：保留的主成分解释的方差比例

# 保留所有主成分
pca = PCA()
X_pca = pca.fit_transform(X)
print("\nPCA转换后的数据：")
print(X_pca)

# 查看各主成分解释的方差比例
print("\n各主成分解释的方差比例：", pca.explained_variance_ratio_)
print("累计解释的方差比例：", pca.explained_variance_ratio_.cumsum())

# 查看各主成分的方差
print("\n各主成分的方差：", pca.explained_variance_)

print('-'*50)

# 指定保留的主成分数量
pca1 = PCA(n_components=2)  # 保留2个主成分
X_pca1 = pca1.fit_transform(X)
print("\n保留2个主成分后的数据：")
print(X_pca1)
print("保留2个主成分解释的方差比例：", pca1.explained_variance_ratio_)
print("累计解释的方差比例：", pca1.explained_variance_ratio_.cumsum())

print('-'*50)

# 指定累计解释方差95%
pca_95 = PCA(n_components=0.95)  # 保留累计解释方差≥95%的主成分
X_pca_95 = pca_95.fit_transform(X)
print("\n使用累计解释方差95%进行PCA降维后的数据：")
print(X_pca_95)
print("降维后的数据形状：", X_pca_95.shape)
print("\n各主成分解释的方差比例：", pca_95.explained_variance_ratio_)
print("累计解释的方差比例：", pca_95.explained_variance_ratio_.cumsum())
print(f"实际保留的主成分数量：{pca_95.n_components_}")



原始数据 X：
[[0 2 0 3]
 [0 1 4 3]
 [0 1 1 3]]
原始数据形状： (3, 4)
--------------------------------------------------

PCA转换后的数据：
[[-1.76504522e+00  3.26860225e-01 -2.98186724e-17]
 [ 2.35339362e+00  1.30744090e-01 -2.98186724e-17]
 [-5.88348405e-01 -4.57604315e-01 -2.98186724e-17]]

各主成分解释的方差比例： [9.64285714e-01 3.57142857e-02 2.85799251e-34]
累计解释的方差比例： [0.96428571 1.         1.        ]

各主成分的方差： [4.50000000e+00 1.66666667e-01 1.33372984e-33]
--------------------------------------------------

保留2个主成分后的数据：
[[-1.76504522  0.32686023]
 [ 2.35339362  0.13074409]
 [-0.58834841 -0.45760432]]
保留2个主成分解释的方差比例： [0.96428571 0.03571429]
累计解释的方差比例： [0.96428571 1.        ]
--------------------------------------------------

使用累计解释方差95%进行PCA降维后的数据：
[[-1.76504522]
 [ 2.35339362]
 [-0.58834841]]
降维后的数据形状： (3, 1)

各主成分解释的方差比例： [0.96428571]
累计解释的方差比例： [0.96428571]
实际保留的主成分数量：1


**各主成分解释的方差比例 vs 累计解释的方差比例**

**区别说明：**

1. **各主成分解释的方差比例** (`explained_variance_ratio_`)
   - 每个主成分**单独**解释的方差比例
   - 例如：[0.96, 0.04, 0.00] 表示：
     - 第1个主成分解释96%的方差
     - 第2个主成分解释4%的方差
     - 第3个主成分解释0%的方差

2. **累计解释的方差比例** (`explained_variance_ratio_.cumsum()`)
   - **从第1个到当前主成分**累计解释的方差比例
   - 例如：[0.96, 1.00, 1.00] 表示：
     - 前1个主成分累计解释96%的方差
     - 前2个主成分累计解释100%的方差
     - 前3个主成分累计解释100%的方差

**实际应用：**
- 查看**各主成分比例**：了解每个主成分的重要性
- 查看**累计比例**：决定保留多少个主成分（通常选择累计比例≥90%或95%）